src/linker/Linker/Driver.cs

```diff
@@ -697,15 +697,16 @@ public bool Run (ILogger customLogger = null)
 
 			try {
 				p.Process (context);
-			} catch (Exception ex) {
-				if (ex is LinkerFatalErrorException lex) {
-					context.LogMessage (lex.MessageContainer);
-					Console.Error.WriteLine (ex.ToString ());
-				} else {
-					context.LogError ($"IL Linker has encountered an unexpected error. Please report the issue at https://github.com/mono/linker/issues \n{ex}", 1012);
-				}
-
+			} catch (LinkerFatalErrorException lex) {
+				context.LogMessage (lex.MessageContainer);
+				Console.Error.WriteLine (lex.ToString ());
```

Comment:
@vitek-karas @mateoatr  Should we do the same thing for `LinkerFatalErrorException`? I don't fully understand why we have this exception type - I thought that its purpose was to carry a `MessageContainer` for non-recoverable but fixable (by the user) errors, but it also has a [ctor](https://github.com/mono/linker/blob/master/src/linker/Linker/LinkerFatalErrorException.cs#L14) that takes a string that suggests it is to be used for non-fixable errors. Why don't we just use standard .NET exception types for internal linker errors?

Reply:
There is a benefit to logging the stacktrace instead of rethrowing here - `LinkerFatalErrorException` will have an exit code of 1, instead of 134 for the unhandled exceptions, allowing us to distinguish based on the exit code (in telemetry for example). This is why I would like to clarify the meaning of `LinkerFatalErrorException`.

Reply:
I also thought that the purpose of `LinkerFatalErrorException` was to carry the `MessageContainer` (we want this so that we can for example specify the origin). But as the name suggests, it's a fatal - non-recoverable error. 

We might want to let it go unhandled, but there are pros and cons:
* Pro - it will trigger Watson and potentially give a chance to get detailed info about the failure through that
* Cons - we might lose the nice formatting - not sure how it will behave with regard to MSBuild (for example, will it get the right origin?)

If we don't rethrow, we can use the message code as the exit code for linker, that way we get nice telemetry. This is probably the best solution - if we swallow it fully we get exit code 1. If we rethrow we get exit code 134 (or whatever), but no details. So catching and using the message code as the exit code gives us most info.

Reply:
Non-recoverable but user-fixable errors should be logged using `LogError`, the `LinkerFatalErrorException` is used to throw whenever the linker ends up in a place where it shouldn't (and the user can do nothing about it). I think there are a few places where we're using this exception without specifying an origin (and we can/should, e.g., [here](https://github.com/mono/linker/blob/master/src/linker/Linker.Steps/MarkStep.cs#L2454)).

I like @vitek-karas idea of catching and using the error code as the exit code to get more informative telemetry about liker failures.

Reply:
I took the suggestion to return the error code - PTAL.